### DISCUSSION TITLE
How to parse a string without quotes and with code_point?

### DISCUSSION THREAD

**Discussion by Life4gal** [foonathan/lexy#159] [Q&A] [answered]
https://github.com/foonathan/lexy/blob/5e2601b3de57e979b5d23b8936cf040eb228fb94/examples/json.cpp#L209-L260

The above example shows how to parse a string with `code_point` and surrounded by `double quotes`, but it can't be used with strings that aren't surrounded by anything.

In this case, `lexy::dsl::identifier` is generally used as an alternative to parsing strings, but there is a problem in that lexy::dsl::identifier only parses `raw` strings, and does not handle code_points.

I think the solution could be to first get a lexeme with lexy::dsl::identifier and then parse the lexeme. But I'm not sure how to do that.

**Comment by foonathan** [comment]
If the string doesn't have quotes, what terminates the string? You can pass those terminators to `dsl::delimited`: https://lexy.foonathan.net/reference/dsl/delimited/#delimited

**Reply by Life4gal** [reply]
I used `dsl::delimited` at first, but my strings start with a non-whitespace character and end with a newline, and I don't think `dsl::delimited` should be able to parse them that way.

**Reply by Life4gal** [reply]
I found that I could use
```cpp
constexpr static auto rule =
		dsl::peek(dsl::unicode::print - dsl::unicode::newline) >>
		dsl::delimited(dsl::unicode::print - dsl::unicode::newline, dsl::unicode::newline)(code_point, escape);
```
To parse a string that is not surrounded by any characters, but this consumes `newline` at the end of the line.

If such a string is bound to be parsed, this is obviously not a problem. But if the string is optional (dsl::opt), then whether newline is consumed is obviously relevant to subsequent parsing. Is there any way to solve this problem? I was thinking of backtracking one character (the newline) after parsing, but I'm not sure how to do that.

**Reply by foonathan** [reply]
You shouldn't need `dsl::peek` before `dsl::delimited`, that is already the behavior it has on its own. If you don't want to consume the newline, you can use `dsl::peek(dsl::unicode::newline)` as the closing terminator.

**Reply by Life4gal** [reply]
```cpp
constexpr static auto rule = dsl::delimited(dsl::peek(dsl::unicode::print - dsl::unicode::newline - dsl::unicode::blank), dsl::peek(dsl::unicode::newline))(code_point, escape) | dsl::error<invalid_string>;

constexpr static auto value = lexy::as_string<std::string>;
```

The above rule solves almost all problems, but it allows empty strings (instead of generating an error), how to reject this?

**Reply by Life4gal** [reply]
https://godbolt.org/z/x56TosjGj

**Reply by foonathan** [reply, marked as answer]
It doesn't allow empty strings? https://lexy.foonathan.net/playground/?id=1rasdbjY3&mode=tree

**Reply by Life4gal** [reply]
https://lexy.foonathan.net/playground/?id=foY5n6s4W&mode=tree

That's what I need! Although I need to manually remove the extra whitespace on the string right side in the callback (I want to allow whitespace inside the string after all)

**Reply by Life4gal** [reply]
It doesn't seem to be fully resolved, it looks fine in the playground.

https://lexy.foonathan.net/playground/?id=aseo8xYGP&mode=tree

But: 
![image](https://github.com/foonathan/lexy/assets/52756109/779c19d6-58fe-41dc-81b1-f944164e4d67)
![image](https://github.com/foonathan/lexy/assets/52756109/62a1ae81-2aa8-4797-ad41-278a0ec282cd)

Why is `tab` considered an illegal character? (BTW, I was using `-control` before)

**Reply by Life4gal** [reply]
![image](https://github.com/foonathan/lexy/assets/52756109/2c348214-ed8b-4b9f-be83-d6f55f264dbf)
:(

**Reply by foonathan** [reply]
You said that your delimited should only contain printable Unicode characters, and TAB just isn't considered to be one according to the Unicode standard. So you'd want `dsl::unicode::print / dsl::lit_c<'\t'>`.

(The playground translates tab to four spaces, which is why it works there.)
